### DISCUSSION TITLE
Add an option to implement SmartPV values to the initial config

### DISCUSSION THREAD

**Discussion by EnerGeh** [dominikamann/oekofen-pellematic-compact#26] [Ideas]
Hi @dominikamann and first of all: Thank you very much for your HACS integration for Ökofen. It works out of the box but I'm missing the values from the SmartPV feature (that features basically a power meter for your power flow direction [buy or sell] and an electric heating element in the water storage tank).
I'd prefer to contribute with a pull request but I'm probably the worst programmer in the world. The values are stored in the $.power domain and are basically 
```
{
 "power":{
  "power_info":"powermeter data",
  "L_usage":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_pv_total":{"val":0, "unit":"W", "factor":1}, 
  "L_pv_1":{"val":0, "unit":"W", "factor":1}, 
  "L_pv_2":{"val":0, "unit":"W", "factor":1}, 
  "L_from_grid":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_to_grid":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_p1":{"val":0 "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_p2":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_p3":{"val":-0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_batt_in":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_batt_out":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_batt_chg":{"val":0, "unit":"%", "factor":1, "min":-32768, "max":32767}, 
  "L_batt_enabled":{"val":0, "format":"0:Inaktiv|1:Aktiv"}, 
  "L_pwr_out_per":{"val":0, "unit":"%", "factor":1, "min":0, "max":100}, 
  "L_pwr_out":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_power2car":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_today_total":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_today_in":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_today_out":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_state":{"val":0, "factor":1}, 
  "L_total":{"val":0, "factor":1}, 
  "L_yesterday":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_total_tyristor":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_yesterday_tyristor":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_total_car":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_yesterday_car":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_total_pv":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_yesterday_pv":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_today_batt_in":{"val":0, "factor":1}, 
  "L_today_batt_out":{"val":0, "factor":1}, 
  "L_yesterday_batt_in":{"val":0, "factor":1}, 
  "L_yesterday_batt_out":{"val":0, "factor":1}, 
  "L_output_mode":{"val":0, "format":"0:Aus|1:Ein"}, 
  "L_range":{"val":0, "unit":"W", "factor":1, "min":500, "max":10000}, 
  "L_offset":{"val":0, "unit":"W", "factor":1, "min":0, "max":10000}, 
  "L_total_in":{"val":0, "factor":1}, 
  "L_total_out":{"val":0, "factor":1}, 
  "L_yesterday_in":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_yesterday_out":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "offtemp":{"val":0, "unit":"°C", "factor":0.1, "min":300, "max":900}
 }
}
```
I'd be very happy to assist with implementing these sensors, but didn't fully understand your code. I think I should add the values to const.py, how do you import that in the configuration during creation? Maybe another switch with "Use SmartPV" would be nice as this seems to be a niche Ökofen add-on that is relatively unusual. Any help is very appreciated for my first commit on github :smiley:

**Comment by dominikamann** [comment]
Hi @EnerGeh,
you are welcome. Thank you for providing all these information. I will implement it within the next few days 👍 If you want you can have a look on the commit. Then you will see how it works. I also use a electric heating element in the water storage tank. I feed with surplus PV energy therefore I use the SolarEdge Smart Heating System. How did you integrated your PV in Ökofen?

**Reply by EnerGeh** [reply]
@dominikamann I'm happy to assist and that you take over the task to further enhance your implementation. I own a nearly 8 year old Pellematic SmartXS, so this may be a bit different from your system but as far as I understand the software is very similar. I recently got an update that enabled the JSON API and was implementing it in HASS with a custom config.yaml but your custom integration is much more efficient and streamlined.
I'd be happy to have a look into your commit when it's ready to be checked :smiley: 
I ordered the official SmartPV add-on a couple of years ago. It features an Ökofen-branded "voltoplus" smart meter that directly interacts with the boiler control system via IP. The heating element is then steered by a thyristor as a phase control modulator and modulates between (I think) 10...2000W electrical.
During the setup there was also an option to integrate a SMA inverter but I didn't opt-in for that as I own the voltoplus meter.

**Comment by dominikamann** [comment]
@EnerGeh can you please provide the full JSON of your Pellematic. This makes it easier for me to test. Thanks.

**Reply by EnerGeh** [reply]
Of course, feel free to ask again if you need any further information. Find attached a somehow anonymized output.
```
{
 "system":{
  "system_info":"system global variables",
  "L_ambient":{"val":105, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_errors":{"val":0, "factor":1, "min":-32768, "max":32767}, 
  "L_usb_stick":{"val":0, "format":"0:Aus|1:Ein"}, 
  "L_existing_boiler":{"val":0, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}
 },
 "weather":{
  "weather_info":"current weather data",
  "L_temp":{"val":70, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_clouds":{"val":100, "unit":"%", "factor":1, "min":-32768, "max":32767}, 
  "L_forecast_temp":{"val":70, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_forecast_clouds":{"val":100, "unit":"%", "factor":1, "min":-32768, "max":32767}, 
  "L_forecast_today":{"val":1, "format":"0:Heute|1:Morgen"}, 
  "L_starttime":{"val":830, "factor":1, "min":-32768, "max":32767}, 
  "L_endtime":{"val":1750, "factor":1, "min":-32768, "max":32767}, 
  "L_source":{"val":"https://www.openweathermap.org", "length":20}, 
  "L_location":{"val":"City", "length":20}, 
  "cloud_limit":{"val":55, "unit":"%", "factor":1, "min":0, "max":100}, 
  "hysteresys":{"val":-40, "unit":"K", "factor":0.1, "min":-200, "max":0}, 
  "offtemp":{"val":-100, "unit":"°C", "factor":0.1, "min":-300, "max":200}, 
  "lead":{"val":120, "unit":"min", "factor":1, "min":0, "max":600}, 
  "refresh":{"val":0}, 
  "oekomode":{"val":1, "format":"0:Aus|1:Ein"}
 },
 "forecast":{
  "forecast_info":"date|temp|cloud|speed|image|code|unit[|sunrise|sunset] code see https://openweathermap.org/weather-conditions",
  "L_w_0":{"val":"", "length":20}, 
  "L_w_1":{"val":"", "length":20}, 
  "L_w_2":{"val":"", "length":20}, 
  "L_w_3":{"val":"", "length":20}, 
  "L_w_4":{"val":"", "length":20}, 
  "L_w_5":{"val":"", "length":20}, 
  "L_w_6":{"val":"", "length":20}, 
  "L_w_7":{"val":"", "length":20}, 
  "L_w_8":{"val":"", "length":20}, 
 },
 "power":{
  "power_info":"powermeter data",
  "L_usage":{"val"0:, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_pv_total":{"val":0, "unit":"W", "factor":1}, 
  "L_pv_1":{"val":0, "unit":"W", "factor":1}, 
  "L_pv_2":{"val":0, "unit":"W", "factor":1}, 
  "L_from_grid":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_to_grid":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_p1":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_p2":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_p3":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_batt_in":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_batt_out":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_batt_chg":{"val":0, "unit":"%", "factor":1, "min":-32768, "max":32767}, 
  "L_batt_enabled":{"val":0, "format":"0:Inaktiv|1:Aktiv"}, 
  "L_pwr_out_per":{"val":0, "unit":"%", "factor":1, "min":0, "max":100}, 
  "L_pwr_out":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_power2car":{"val":0, "unit":"W", "factor":1, "min":-32768, "max":32767}, 
  "L_today_total":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_today_in":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_today_out":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_state":{"val":2, "factor":1}, 
  "L_total":{"val":0, "factor":1}, 
  "L_yesterday":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_total_tyristor":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_yesterday_tyristor":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_total_car":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_yesterday_car":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_total_pv":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_yesterday_pv":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_today_batt_in":{"val":0, "factor":1}, 
  "L_today_batt_out":{"val":0, "factor":1}, 
  "L_yesterday_batt_in":{"val":0, "factor":1}, 
  "L_yesterday_batt_out":{"val":0, "factor":1}, 
  "L_output_mode":{"val":1, "format":"0:Aus|1:Ein"}, 
  "L_range":{"val":0, "unit":"W", "factor":1, "min":500, "max":10000}, 
  "L_offset":{"val":0, "unit":"W", "factor":1, "min":0, "max":10000}, 
  "L_total_in":{"val":0, "factor":1}, 
  "L_total_out":{"val":0, "factor":1}, 
  "L_yesterday_in":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "L_yesterday_out":{"val":0, "unit":"kWh", "factor":1.0E-4}, 
  "offtemp":{"val":0, "unit":"°C", "factor":0.1, "min":300, "max":900}
 },
 "hk1":{
  "hk_info":"heating circuit data",
  "L_roomtemp_act":{"val":0, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_roomtemp_set":{"val":80, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_flowtemp_act":{"val":448, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_flowtemp_set":{"val":80, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_comfort":{"val":0, "unit":"K", "factor":0.1, "min":-32768, "max":32767}, 
  "L_state":{"val":2, "factor":1}, 
  "L_statetext":"Warmwasser Vorrang aktiv",
  "L_pump":{"val":0, "format":"0:Aus|1:Ein"}, 
  "remote_override":{"val":0, "unit":"K", "factor":0.1, "min":-32768, "max":32767}, 
  "mode_auto":{"val":1, "format":"0:Aus|1:Auto|2:Heizen|3:Absenken"}, 
  "time_prg":{"val":0, "format":"0:Zeit 1|1:Zeit 2"}, 
  "temp_setback":{"val":180, "unit":"°C", "factor":0.1, "min":100, "max":400}, 
  "temp_heat":{"val":210, "unit":"°C", "factor":0.1, "min":100, "max":400}, 
  "temp_vacation":{"val":150, "unit":"°C", "factor":0.1, "min":100, "max":400}, 
  "name":{"val":"name1", "length":20}, 
  "oekomode":{"val":0, "format":"0:Aus|1:Komfort|2:Minimum|3:Ökologisch"}, 
  "autocomfort":{"val":-1, "format":"0:Aus|1:Ein|2:Morgens|3:Abends"}, 
  "autocomfort_sunset":{"val":0, "unit":"min", "factor":1, "min":-120, "max":120}, 
  "autocomfort_sunrise":{"val":0, "unit":"min", "factor":1, "min":-120, "max":120}
 },
 "hk2":{
  "hk_info":"heating circuit data",
  "L_roomtemp_act":{"val":0, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_roomtemp_set":{"val":80, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_flowtemp_act":{"val":315, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_flowtemp_set":{"val":80, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_comfort":{"val":0, "unit":"K", "factor":0.1, "min":-32768, "max":32767}, 
  "L_state":{"val":2, "factor":1}, 
  "L_statetext":"Warmwasser Vorrang aktiv",
  "L_pump":{"val":0, "format":"0:Aus|1:Ein"}, 
  "remote_override":{"val":0, "unit":"K", "factor":0.1, "min":-32768, "max":32767}, 
  "mode_auto":{"val":1, "format":"0:Aus|1:Auto|2:Heizen|3:Absenken"}, 
  "time_prg":{"val":0, "format":"0:Zeit 1|1:Zeit 2"}, 
  "temp_setback":{"val":180, "unit":"°C", "factor":0.1, "min":100, "max":400}, 
  "temp_heat":{"val":205, "unit":"°C", "factor":0.1, "min":100, "max":400}, 
  "temp_vacation":{"val":150, "unit":"°C", "factor":0.1, "min":100, "max":400}, 
  "name":{"val":"name2", "length":20}, 
  "oekomode":{"val":0, "format":"0:Aus|1:Komfort|2:Minimum|3:Ökologisch"}, 
  "autocomfort":{"val":-1, "format":"0:Aus|1:Ein|2:Morgens|3:Abends"}, 
  "autocomfort_sunset":{"val":0, "unit":"min", "factor":1, "min":-120, "max":120}, 
  "autocomfort_sunrise":{"val":0, "unit":"min", "factor":1, "min":-120, "max":120}
 },
 "ww1":{
  "ww_info":"domestic hot water data",
  "L_temp_set":{"val":750, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_ontemp_act":{"val":721, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_offtemp_act":{"val":721, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_pump":{"val":0, "format":"0:Aus|1:Ein"}, 
  "L_state":{"val":16912, "factor":1}, 
  "L_statetext":"Zeit innerhalb Zeitprogramm|Quellentemp unterhalb Warmwassertemp|Anforderung Ein",
  "time_prg":{"val":0, "format":"0:Zeit 1|1:Zeit 2"}, 
  "sensor_on":{"val":0, "format":"0:WW|1:TPO|2:TPM|3:SpUnten"}, 
  "sensor_off":{"val":0, "format":"0:WW|1:TPO|2:TPM|3:SpUnten"}, 
  "mode_auto":{"val":1, "format":"0:Aus|1:Auto|2:Ein"}, 
  "mode_dhw":{"val":1, "format":"0:Aus|1:Auto|2:Ein"}, 
  "heat_once":{"val":0, "format":"0:Aus|1:Ein"}, 
  "temp_min_set":{"val":450, "unit":"°C", "factor":0.1, "min":80, "max":800}, 
  "temp_max_set":{"val":750, "unit":"°C", "factor":0.1, "min":80, "max":800}, 
  "name":{"val":"", "length":20}, 
  "smartstart":{"val":60, "unit":"min", "factor":1, "min":0, "max":90}, 
  "use_boiler_heat":{"val":1, "format":"0:Aus|1:Ein"}, 
  "oekomode":{"val":3, "format":"0:Aus|1:Komfort|2:Minimum|3:Ökologisch"}
 },
 "pe1":{
  "pe_info":"pellematic data",
  "L_temp_act":{"val":678, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_temp_set":{"val":800, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_ext_temp":{"val":-32768, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_frt_temp_act":{"val":7344, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_frt_temp_set":{"val":7081, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_frt_temp_end":{"val":7600, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_br":{"val":0, "format":"0:Aus|1:Ein"}, 
  "L_ak":{"val":0, "format":"0:Aus|1:Ein"}, 
  "L_not":{"val":1, "format":"0:Aus|1:Ein"}, 
  "L_stb":{"val":1, "format":"0:Aus|1:Ein"}, 
  "L_modulation":{"val":100, "unit":"%", "factor":1, "min":-32768, "max":32767}, 
  "L_runtimeburner":{"val":5056, "unit":"zs", "factor":0.01}, 
  "L_resttimeburner":{"val":8000, "unit":"zs", "factor":0.01}, 
  "L_currentairflow":{"val":0, "unit":"%", "factor":1, "min":-32768, "max":32767}, 
  "L_lowpressure":{"val":710, "unit":"EH", "factor":0.1, "min":-32768, "max":32767}, 
  "L_lowpressure_set":{"val":710, "unit":"EH", "factor":0.1, "min":-32768, "max":32767}, 
  "L_fluegas":{"val":62, "unit":"%", "factor":1, "min":-32768, "max":32767}, 
  "L_uw_speed":{"val":100, "unit":"%", "factor":1, "min":-32768, "max":32767}, 
  "L_state":{"val":4, "format":"0:Dauerlauf|1:Start|2:Zuendung|3:Softstart|4:Leistungsbrand|5:Nachlauf|6:Aus|7:Saugen|8:! Asche !|9:! Pellets !|10:Pell Switch|11:Störung|12:Einmessen|13:1|14:1|15:1|16:1|17:1|18:1|19:1|20:1|21:1|22:1|23:1|24:1|25:1|26:1|27:1|28:1|29:1|30:1|31:1|32:1|33:1|34:1|35:1|36:1|37:1|38:1|39:1|40:1|41:1|42:1|43:1|44:1|45:1|46:1|47:1|48:1|49:1|50:1|51:1|52:1|53:1|54:1|55:1|56:1|57:1|58:1|59:1|60:1|61:1|62:1|63:1|64:1|65:1|66:1|67:1|68:1|69:1|70:1|71:1|72:1|73:1|74:1|75:1|76:1|77:1|78:1|79:1|80:1|81:1|82:1|83:1|84:1|85:1|86:1|87:1|88:1|89:1|90:1|91:1|92:1|93:1|94:1|95:1|96:1|97:Aus|98:Aus|99:Aus|100:Aus|101:Aus"}, 
  "L_statetext":"Leistungsbrand",
  "L_type":{"val":7, "format":"0:PE|1:PES|2:PEK|3:PESK|4:SMART V1|5:SMART V2|6:CONDENS|7:SMART XS|8:SMART V3|9:COMPACT|10:AIR"}, 
  "L_starts":{"val":18144, "factor":1}, 
  "L_runtime":{"val":14206, "unit":"h", "factor":1}, 
  "L_avg_runtime":{"val":46, "unit":"min", "factor":1}, 
  "L_uw_release":{"val":350, "unit":"°C", "factor":0.1, "min":-32768, "max":32767}, 
  "L_uw":{"val":100, "unit":"%", "factor":1, "min":-32768, "max":32767}, 
  "L_storage_fill":{"val":3949, "unit":"kg", "factor":1}, 
  "L_storage_min":{"val":400, "unit":"kg", "factor":1, "min":0, "max":4000}, 
  "L_storage_max":{"val":6000, "unit":"kg", "factor":1, "min":150, "max":30000}, 
  "L_storage_popper":{"val":22, "unit":"kg", "factor":1, "min":-32768, "max":32767}, 
  "storage_fill_today":{"val":4, "unit":"kg", "factor":1, "min":-32768, "max":32767}, 
  "storage_fill_yesterday":{"val":11, "unit":"kg", "factor":1, "min":-32768, "max":32767}, 
  "mode":{"val":1, "format":"0:Aus|1:Auto|2:Ein"}
 },
   "error":{
   }
}

```

**Comment by dominikamann** [comment]
@EnerGeh are you sure the current integration works for you? Your JSON has a completely different format. If i test it with my code you dont get really nice values 😄. Which version do you use on your pellematic heater?

Default:
![grafik](https://user-images.githubusercontent.com/29973737/232112102-0f5f1d19-cbc0-402f-9dbb-862672090fae.png)

Yours:
![grafik](https://user-images.githubusercontent.com/29973737/232112211-5c460aaa-c047-499c-ab42-b5b6191d7546.png)

**Reply by EnerGeh** [reply]
I feel so sorry! I thought it might be a good idea to send you the detailed output from the JSON API as it is a kind of a description. You can get this output when you add a `?` at the end like `http://[ip]:4321/[password]/all?`. When you omit that character, you will get the values as shown in your screenshot.
Please take my apologies for that mistake!

**Reply by EnerGeh** [reply]
Additional to my findings above, I found some differences between my Pellematic SmartXS and your Pellematic Compact. You're referencing a hot water circuit `hwX` in the setup, in my case, this is a domestic water circuit in the `ww1` register that is already recognized by v3.0.0.
But I have unknown variables that have no representative in the API:
![grafik](https://user-images.githubusercontent.com/28715981/232225469-c15f11e8-6b47-4e20-b9f3-52e13210cf8d.png)
Is it hard to omit these sensors when they don't exist?

**Reply by dominikamann** [reply]
Hi @EnerGeh,

thanks for the hint with "?". I was not aware of this feature 😅. Helps a lot.

I will add a configuration for you. Then the integration will not add the sensors in your case. 

The errors (1..5) are fine. They will have shown an error text once a error occurs on your heater. The sensor must be always present that it is possible to use them in e.g automations.

**Reply by EnerGeh** [reply]
@dominikamann Okay, it's pretty on top in the description when you open the API without any command like this: http://[IP]:4321/[Password]/

**Comment by dominikamann** [comment]
@EnerGeh  I have implemented the new JSON format & SmartPV support.

You have not included any data in the power section. Every value is 0, so it is difficult to test this. Please test the feature and report back if everything works. A screenshot would also be nice.

Thank you.

See release https://github.com/dominikamann/oekofen-pellematic-compact/releases/tag/v3.0.0

**Reply by EnerGeh** [reply]
And once again a huge thank you for your fast reply and the implementation time world record. It works like a charm as far as I can see for now. I added a screenshot below as you asked for some real world numbers.
One thing I observed is that the scaling of the values `Power today in`, `Power today out` and `Power today total` is incorrect. According to the docs, they should be divided by 10000 (multiplied with 10⁻⁴).
![grafik](https://user-images.githubusercontent.com/28715981/232200126-82bdb645-3d55-4557-8e1d-986dba86762e.png)

**Reply by EnerGeh** [reply]
I found a couple of sensor where the scaling doesn't match as well. They are for your reference:
```
"L_today_total":{"val":-19255000, "unit":"kWh", "factor":1.0E-4}
"L_today_in":{"val":64000, "unit":"kWh", "factor":1.0E-4}
"L_today_out":{"val":334000, "unit":"kWh", "factor":1.0E-4}
"L_yesterday":{"val":-18985000, "unit":"kWh", "factor":1.0E-4}
"L_total_tyristor":{"val":14273825, "unit":"kWh", "factor":1.0E-4}
"L_yesterday_tyristor":{"val":227297, "unit":"kWh", "factor":1.0E-4}
"L_total_car":{"val":0, "unit":"kWh", "factor":1.0E-4}
"L_yesterday_car":{"val":0, "unit":"kWh", "factor":1.0E-4}
"L_total_pv":{"val":0, "unit":"kWh", "factor":1.0E-4}
"L_yesterday_pv":{"val":0, "unit":"kWh", "factor":1.0E-4}
"L_yesterday_in":{"val":151000, "unit":"kWh", "factor":1.0E-4}
"L_yesterday_out":{"val":403000, "unit":"kWh", "factor":1.0E-4}
```

**Reply by dominikamann** [reply]
@EnerGeh You are welcome! Thanks for the feedback I will correct/implement the scalings 👍

**Comment by EnerGeh** [comment]
And just for your information: When you try to use the newly created sensors in the energy dashboard, you'll get a warning that the `last_reset` state is missing:
![grafik](https://user-images.githubusercontent.com/28715981/232200402-2782ee13-6725-4462-90b9-402653703f77.png)
I don't know if it could be easily added, a workaround or maybe another solution could be to manually create a virtual sensor where the `last_reset` state will be set to 12pm every day.

**Reply by EnerGeh** [reply]
I investigated that further and I think the issue here is, that the "net usage" and "to the net" values are not meant to be set by daily sensors without that `last_reset` attribute. I think we can take the `Power total in` and `Power total out`for that but that needs to be configured in kWh accordingly.

**Reply by dominikamann** [reply]
I will check that.

**Comment by EnerGeh** [comment]
So, my last comment for today: Please take my apologies for that huge input in different posts today. I tried to give it a structure but now it looks very messy to me. If you ever come to the region of Südpfalz, you will get a free or two (or ten) :blush:

**Reply by dominikamann** [reply]
@EnerGeh überhaupt kein Problem 😉 Ich schaue mir die Fehler an. Top das schon mal Werte ankommen 👍 Immer etwas nervig wenn man es selber an seiner Anlage nicht testen kann. Ich gehe davon aus Du sprichst von "Bier" 🍺. Falls ich mal in der Gegend sein sollte melde ich mich 💪😅

**Reply by EnerGeh** [reply]
Ach verflixt, das kommt davon, wenn man ständig seine eigenen Sätze korrigiert, dann geht hier und da auch mal ein Wort verloren :beer: .
Und nochmals vielen lieben Dank, dass du dir hier so Mühe gibst - deine Custom Integration ist echt Gold wert!

**Comment by dominikamann** [comment]
@EnerGeh I have released version 3.0.1. Can you please check your sensor values & if you can add the sensors to your energy dashboard now? I've had a stressful few days .... sorry for making you wait so long ...

https://github.com/dominikamann/oekofen-pellematic-compact/releases/tag/v3.0.1

**Reply by EnerGeh** [reply]
@dominikamann Thank you very much for your continued support in that case and please don't worry at all! Sometimes (often) real life slows down open source software and this is why life is complicated :smiling_face_with_tear:
The values of the sensors didn't change scaling and as far is I understand the commit you referenced in the release notes this is because you scale the Kilowatt values but you should scale the Kilowatthours, shouldn't you?
And for adding my admiration: I like the way you apply the scaling, I would have done that with a simple division by ten to every single value but you iterate over the units of measurement which seems to be a much more elegant solution. Chapeau! :smile:

**Reply by dominikamann** [reply]
Sorry for the try & error. I already changed it but did not checked it in. Can you try 3.0.2? Thanks.

**Reply by EnerGeh** [reply]
No need to excuse, I'm fine with testing and willing to help :+1: 
The scaling of the values is now correct so far.
I just compared the values `L_total_in` and `L_total_out` with the Touch Display and these values also represent energy, so they should have the kWh unit but without scaling. I don't know why they are not correctly represented in the `/all?` description.

One last question: As the `L_total_in` and `L_total_out` values are not very precise in terms of the values, would you mind adding the `last_reset` state to the sensors `L_today_in` and `L_today_out` so that I can use them in the Energy Dashboard?

**Reply by dominikamann** [reply]
So L_total_in & L_total_out is kW and does not need scaling - right?
L_today_in & L_today_out should already work in the energy dashboard as I introduced the status class "Total increasing" (https://developers.home-assistant.io/blog/2021/08/16/state_class_total/) for these sensors. Maybe you need to go to "Developer Tools" and execute problem solving for the entities because the state class changed (https://community.home-assistant.io/t/energy-dashboard-sensor-not-showing-up/438537/3).

**Reply by EnerGeh** [reply]
Partially - `L_total_in` and `L_total_out` is in `kWh` (Energy) and **not kW** (Power) and can therefore be used in the Energy Dashboard.
With your hint I found out that I didn't record all sensors, so these Sensors didn't appear under the Developer Tools - Statistics - I changed that recently and will have a look tomorrow as I think this will need some time to generate data.

Additionally, I'll try to generate a kind of a translation table to support you much better with real life values and compare that with the according fields from the Touch Display.

Another side-question: Do you know if local access to the Display is still possible with v4.00b? When I try to access the display's IP address via plain http (port 80) , I'm redirected to https://my.oekofen.info and have to register my Email address. This was not necessary in the previous software version. (I don't want to use any cloud service from Ökofen)

**Reply by dominikamann** [reply]
@EnerGeh any news?

Regarding your side question I only know the online / cloud portal. May @lucaferruci knows more?

**Reply by EnerGeh** [reply]
@dominikamann I feel so Sorry, real life hits very hard currently. I'll do my best to further investigate that  as soon as possible.

**Comment by lucaferrucci** [comment]
the portal redirects to the machine's IP, but I don't know if it works normally by disconnecting the internet via LAN when you log in, just try


> Il giorno 5 mag 2023, alle ore 10:46, Dominik Amann ***@***.***> ha scritto:
> 
> Regarding your side question I only know the online / cloud portal. May @lucaferruci knows more?

**Comment by lucaferrucci** [comment]
Hi, I hadn't read your conversation in full, I would like an idea on my problem that I would like to solve with photovoltaics, I have an off-grid photovoltaic system with a Voltronic Axpert 6kW inverter, but I would also like to use the electrical resistance to exploit the excess energy.
The smrtPv okofen or ohmpilot fronius device uses the energy transferred by the system to the manager to power the electrical resistance with the 0-10v Pwm signal.
In my case I don't know how to measure the available energy and power the resistance, my inverter generates energy when I absorb it, so I have no way of measuring what I would have available, for the resistance I can use a shelly 1pm relay and a dimmer always shelly which has 0-10v output, so this pair would be perfect for using, but I can't figure out how to automate it.
Good question, but I've been thinking about how I could do it for some time, thank you all for the valuable work you've done

**Comment by level420** [comment]
@EnerGeh sorry for kicking in here after more than 2 years: 
Do you track the power of the smart PV used for heating?

**Reply by dominikamann** [reply]
@level420 Is the data available on the JSON interface? Then this could be implemented.

**Reply by acid-at** [reply]
Hello and thank you for your great Integration to home assistant. I use it with pleasure! I use also the smart pv function, but i missing the control function for smart pv. Isn't possible or do you missing just the actual json api documentation for 4.02b version?

**Reply by dominikamann** [reply]
Hi @acid-at,
so read is implemented I guess. I can not test it as I do not have any Oekofen Smart PV Installation at Home. What do you expect to control?

**Reply by level420** [reply]
@dominikamann Sorry for being late with my answer. I was in vacations for a while. I've got it working with the current version of the integration. See https://github.com/dominikamann/oekofen-pellematic-compact/discussions/127 for the answer.

**Reply by level420** [reply]
@acid-at I only need the read functionality because I want to integrate it in the house energy flow view.

**Reply by acid-at** [reply]
Hi @dominikamann. Sorry my delay, i was very busy the last time.
I search for 2 more options:
- Smart PV: Control electric heater to turn the function on/off
- Compact: "Ash amount PE" to see the actual ash amount and reset if neccesary

Would that be possible?

<img width="692" height="460" alt="Ökofen_Ash_amount" src="https://github.com/user-attachments/assets/19ab251a-fc0b-4345-b4a2-fe7bd329af31" />
<img width="691" height="291" alt="Ökofen_Smart_pv" src="https://github.com/user-attachments/assets/bf08a11a-3288-49e8-9219-541f23f0ae91" />
